### ISSUE TITLE
alarm proposal: detecting port scans with netdata

### ISSUE THREAD

**Issue by ktsaou** [firehol/netdata#970]
On the attacker: `nmap -sS TARGET_IP` or `nmap -sT TARGET_IP`.
On the target netdata:

IPv4:
![image](https://cloud.githubusercontent.com/assets/2662304/18607757/daa3253c-7cde-11e6-91e6-2d7ed4750d5a.png)

The spike in `OutRsts` is a very good indication that hundreds of attempts are made to connect to non-existing ports. This metric could also be used to detect that a busy server has crashed...

Connection Tracker (the firewall on this side is `policy accept`, however this is interesting):
![image](https://cloud.githubusercontent.com/assets/2662304/18607770/0be47bc8-7cdf-11e6-8d2f-552dfbe0d5be.png)

`insert` and `delete_list` match exactly what the attacker did. Hm...

If there are enough daemons listening, there will also be a utilization spike, like this:

Applications CPU:
![image](https://cloud.githubusercontent.com/assets/2662304/18607773/5595fc88-7cdf-11e6-8ab2-4ced9d3f4f84.png)

It is system CPU actually:
![image](https://cloud.githubusercontent.com/assets/2662304/18607774/719bc7be-7cdf-11e6-96dc-e798c6454743.png)

Application Minor Page Faults (most applications fork to serve the requests):
![image](https://cloud.githubusercontent.com/assets/2662304/18607780/9b94f8b0-7cdf-11e6-8ccd-794fb2f1d33c.png)

**Comment by ktsaou** [comment]
And I just realized netdata does not collect `icmp` data from `/proc/net/snmp` and `/proc/net/snmp6`....

**Comment by ktsaou** [comment]
Added IPv4 ICMP and UDPLite with PR #973. No difference though. The port scan is rejected with TCP RST, so no ICMP messages are sent back.

**Comment by ktsaou** [comment]
ok, here it is:

```
   alarm: 1m_ipv4_tcp_resets
      on: ipv4.tcphandshake
  lookup: average -1m at -10s unaligned absolute of OutRsts
   units: tcp resets/s
   every: 10s
    info: average TCP RESETS this host is sending, over the last minute

   alarm: 10s_ipv4_tcp_resets
      on: ipv4.tcphandshake
  lookup: average -10s unaligned absolute of OutRsts
   units: tcp resets/s
   every: 10s
    warn: $this > ($1m_ipv4_tcp_resets * (($status >= $WARNING)  ? (1.5) : (3)))
    crit: $this > ($1m_ipv4_tcp_resets * (($status == $CRITICAL) ? (1.5) : (5)))
   delay: up 0 down 30m multiplier 1.5 max 1h
    info: average TCP RESETS this host is sending, over the last 10 seconds (this can be an indication that a port scan is made, or that a service running on this host has crashed)
      to: sysadmin
```

This is how it works:

1. The `1m_ipv4_tcp_resets` is used just as a variable. It resolves to average TCP RESETS over the last minute (actually one minute past 10 seconds ago). We use this a baseline, a reference, of how the machine goes over normal conditions.

2. Then the alarm `10s_ipv4_tcp_resets` calculates the same average over the last 10 seconds. If this number, the average TCP reset over the last 10 seconds is higher than 3 time the average of the last minute, a warning is issued. If it is 5 times bigger, a critical notification is triggered.

Of course we need to understand if this a valid scenario. Is it possible the TCP resets a host is sending to be increased 3x, 4x or even 5x times over a few seconds under normal conditions?

An nmap on this host, triggers an alarm:

slack:

![image](https://cloud.githubusercontent.com/assets/2662304/18609086/d0a6ff18-7d01-11e6-9d65-41eebcfb6407.png)

email:

![image](https://cloud.githubusercontent.com/assets/2662304/18609090/f0463bb8-7d01-11e6-8012-681b6262f1b0.png)

**Comment by ktsaou** [comment]
To avoid false positives, I will merge this with warning condition only matching 4 times the baseline.

**Comment by qooob** [comment]
@ktsaou Sorry for a little bit of topic, I have just received such email as well. But my question is why `View Netdata` blue button contains links to the external domain (_registry.my-netdata.io_). Is there a way to change this in `/etc/netdata`? Why is this happening anyway?

**Comment by ktsaou** [comment]
of course: https://github.com/firehol/netdata/wiki/mynetdata-menu-item

**Comment by oviliz** [comment]
Hi guys, it looks like the ipv4.tcphandshake doesn't send the confirmation when it clears (I do use Slack to collect the reports). I was investigating one report to then find out that all warning reports are shown as clear in the GUI but we receive no confirmation.

**Comment by ktsaou** [comment]
@oviliz which one? [this](https://github.com/firehol/netdata/blob/eb551d2c19c2fbbc5d702fd8d01f4c1e78b4f990/conf.d/health.d/tcp_resets.conf#L3-L12)?

You can see in that file that there are other alarms with [this option](https://github.com/firehol/netdata/blob/eb551d2c19c2fbbc5d702fd8d01f4c1e78b4f990/conf.d/health.d/tcp_resets.conf#L31), that disables the clear notification. Unfortunately, the way these alarms work, without this option, you will receive a clear notification even if the problem persists.

The reason is that such alarm is comparing 2 durations. Example: the average of the last 10 seconds vs the average of the last minute. In less than a minute after being raised, the comparison will most probably find that the difference is not that high to keep the alarm raised. So, it will send a clear notification, even if the problem remains.

I understand the description should state that, to avoid confusion...

**Comment by ktsaou** [comment]
ok, I added a note to the description of all those alarms.
It will be merged with #2121

**Comment by ktsaou** [comment]
merged it.
